src/uct/tcp/tcp_ep.c

```diff
@@ -1992,23 +2000,21 @@ ucs_status_t uct_tcp_ep_put_zcopy(uct_ep_h uct_ep, const uct_iov_t *iov,
     ctx->super.length = sizeof(put_req);
     put_req.addr      = remote_addr;
     put_req.length    = ep->tx.length;
-    put_req.sn        = ep->tx.put_sn + 1;
+    put_req.sn        = ep->tx.sn;
 
     status = uct_tcp_ep_am_sendv(ep, 0, &ctx->super, UCT_TCP_EP_PUT_ZCOPY_MAX,
                                  &put_req, ctx->iov, ctx->iov_cnt);
     if (ucs_unlikely(status != UCS_OK)) {
         return status;
     }
 
-    ep->tx.put_sn++;
-
-    if (!(ep->flags & UCT_TCP_EP_FLAG_PUT_TX_WAITING_ACK)) {
-        /* Add UCT_TCP_EP_FLAG_PUT_TX_WAITING_ACK flag and increment iface
-         * outstanding operations counter in order to ensure returning
-         * UCS_INPROGRESS from flush functions and do progressing.
-         * UCT_TCP_EP_FLAG_PUT_TX_WAITING_ACK flag has to be removed upon PUT
-         * ACK message receiving if there are no other PUT operations in-flight */
-        ep->flags |= UCT_TCP_EP_FLAG_PUT_TX_WAITING_ACK;
+    ucs_assert(ep->put_cnt != UINT32_MAX);
```

Comment:
maybe better return `NO_RESOURCE` instead?

Reply:
why? we could do several PUT operations simultaneously

Reply:
i mean instead of having an assert you can return `NO_RESOURCE`, until some puts confirmed

Reply:
good idea, done